examples/trusted_information_retrieval/module_1/rust/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+//! scheme is implemented:
+//! - The module receives a request for a specific database entry specified by key.
+//! - Starts continuously requesting chunks of data from the database.
+//! - Processes them and finds the requested entry.
+//! - Sends a response only when all chunks have been requested from the database (even though it
+//!   might have found the requested value earlier).
+//!
+//! This is necessary so we will not leak information about a particular chunk that contains the
+//! requested entry, since an attacker could analyse network traffic.
+
+pub mod proto {
+    include!(concat!(env!("OUT_DIR"), "/oak.examples.database.rs"));
+}
+
+use anyhow::{anyhow, Context};
+use oak::grpc;
+use oak_abi::label::Label;
+pub use proto::DatabaseProxyClient;
+use proto::{
+    DatabaseClient, DatabaseEntry, DatabaseProxy, DatabaseProxyDispatcher, GetDatabaseEntryRequest,
+    GetDatabaseEntryResponse, ListDatabaseEntriesRequest,
+};
+
+/// Expected number of database entries per request.
+const DATABASE_PAGE_SIZE: u32 = 5;
+
+pub fn get_database_entry(id: &str, database_url: &str) -> grpc::Result<DatabaseEntry> {
+    let client = oak::grpc::client::Client::new_with_label(
+        &oak::node_config::wasm("database_proxy", "database_proxy_main"),
+        &Label::public_untrusted(),
+    )
+    .map(DatabaseProxyClient)
+    .ok_or_else(|| {
+        grpc::build_status(
+            grpc::Code::Unavailable,
+            "Couldn't create a database proxy Node",
+        )
+    })?;
+
+    let request = GetDatabaseEntryRequest {
+        key: id.to_string(),
+        database_url: database_url.to_string(),
+    };
+    client.get_database_entry(request).and_then(|response| {
+        response
+            .entry
+            .ok_or_else(|| grpc::build_status(grpc::Code::NotFound, "Empty response"))
+    })
+}
+
+/// Oak Node that connects to an external database.
+#[derive(Default)]
+pub struct DatabaseProxyNode;
+
+impl DatabaseProxyNode {
+    /// Create a gRPC client pseudo-Node.
+    fn get_grpc_client(database_url: &str) -> anyhow::Result<DatabaseClient> {
+        oak::grpc::client::Client::new_with_label(
+            &oak::node_config::grpc_client(database_url),
+            &Label::public_untrusted(),
+        )
+        .map(DatabaseClient)
+        .context("Couldn't create a gRPC client")
+    }
+
+    /// Load a database subset defined by `offset` and the number of requested elements
+    /// (`page_size`).
+    fn list_database_entries(
+        offset: u32,
+        page_size: u32,
+        database_url: &str,
+    ) -> anyhow::Result<Vec<DatabaseEntry>> {
+        let request = ListDatabaseEntriesRequest {
+            offset: offset as i32,
+            page_size: page_size as i32,
+        };
+        Self::get_grpc_client(database_url)?
+            .list_database_entries(request)
+            .map(|response| response.entries)
+            .map_err(|error| anyhow!("gRPC send error: {:?}", error))
```

Comment:
does `.context` not work here?

Reply:
I think so. it would require reimplementing this for `rpc::Status`:
https://github.com/dtolnay/anyhow/blob/55343b9d6cb8fe12486b916fcc1e1588dc4cef11/src/context.rs#L42-L60